### PR TITLE
627 add custom class field to categories

### LINKED ISSUE #627: Add Custom class Field to categories
Add Custom Class Field to categories - give possibility to add css classes to categories, same like in Interactive forms.

### PR DESCRIPTION
**Pull request by BeliaevFiodor**
Resolves #627  
Type: feature

## Issue
it is ready

## Breaking changes
custom class field to categories added

## Testing
open a category in admin page. So it is custom class fied appears at bottom of list

### CODE REVIEW

**Comment by PatrykPorabik** [comment]
Hi,
I'm not able to merge this PR with the last change that you've made in "Fix" commit. Why have you removed the Settings.txt from the file?

**Comment by BeliaevFiodor** [comment]
Hi.
I needed to reinstall DB and it was not possible without it. Now it is fixed

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

пн, 14 окт. 2019 г. в 14:39, Patryk Porabik <notifications@github.com>:

**Comment by PatrykPorabik** [comment]
settings.txt is a crucial file for store, it can be published on GitHub. I meant about change in commit file where you delete file from excluding, not about adding it to PR. Just restore the removed part of this file and send PR again.

**Comment by BeliaevFiodor** [comment]
I'm sorry for this delay

**Comment by PatrykPorabik** [comment]
Hi Fiodor, it's still not about that change. Please check the default Grand.Web.csproj and make your file the same. Send us just changes listed in your first commit.

**Comment by BeliaevFiodor** [comment]
I just get Grand.Web.csproj from 1-st commit. is it enough?

**Comment by PatrykPorabik** [comment]
Fiodor, what about frontend? If we add class in admin panel, what will happen? It should add this class to created category. This class should be used when user would like to add some action to it. For example create mega menu without additional plugins.

File: Grand.Web/App_Data/InstalledPlugins.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+DiscountRequirements.Standard
```
Comment (PatrykPorabik) [inline Grand.Web/App_Data/InstalledPlugins.txt]: Please remove this whole file from commit.

File: Grand.Web/App_Data/Settings.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+DataProvider: mongodb
```
Comment (PatrykPorabik) [inline Grand.Web/App_Data/Settings.txt]: That file also should be removed from the commit, as I wrote few times.

File: Grand.Web/Grand.Web.csproj [hunk 3]
```diff
@@ -42,4 +42,4 @@
     <None Remove="Plugins\**" />
   </ItemGroup>
   
-</Project>
+</Project>
```
Comment (PatrykPorabik) [inline Grand.Web/Grand.Web.csproj:45]: That change also is unnecessary.

**Comment by BeliaevFiodor** [comment]
Hello.
I just had pull a new commit. Class field is receiving in frontend now. But I afraid I can lost some models, where this field is necessary. And I didn't undestand where I should use it. In a case of CustomerAction Class field using while the full html string returning to controller. But Category is drowing in some kind of widget, which I didn't found. 
And I have some ideas about css. If we should use it not only in categories and in customer actions, perhaps the better way is to store styles and classes in separate tables in DB? It allow us to create such kind of styles and classes library to make interface more frendly?

**Review by PatrykPorabik — CHANGES_REQUESTED** [review]
I've left comments in the files which should not appear in commits. In general in pull request, not only in this one, including Settings.txt file is dangerous, because you share your sensitive data with audience. Then, if I try to install GrandNode with this PR included, I won't be able to run it. I will receive the error, because I don't have this db created locally. If you don't know how to remove it or ignore it during the sending PR, please ask @dborovsky for help, he always does it correctly. 

In the case of frontend - this class should be added to the class=" " attribute of each hyperlink of categories. This is enough. There is no need to create additional tables. And it's just about categories. It just should appear here:
image

**Comment by BeliaevFiodor** [comment]
it seems to be done.

**Comment by BeliaevFiodor** [comment]
Hi. I think i did it. But can not turn off red label at "change requested". But it looks like ready

File: Grand.Web/App_Data/Localization/Upgrade/EN_410_420.nopres.xml [hunk 4]
```diff
@@ -449,6 +449,12 @@
   <LocaleResource Name="Admin.Catalog.Categories.Fields.HideOnCatalog.hint">
     <Value>Allow store owner to hide some categories on category page (subcategories before the list of products).</Value>
   </LocaleResource>
+  <LocaleResource Name="Admin.Catalog.Categories.Fields.Class">
```
Comment (PatrykPorabik) [inline Grand.Web/App_Data/Localization/Upgrade/EN_410_420.nopres.xml]: You've added this resources to the wrong file. It's file related for upgrade from 4.10 to 4.20 version of GrandNode. You should add this code to the upgrade file 4.50 to 4.60.

**Comment by BeliaevFiodor** [comment]
ok. I did it

File: Grand.Web/App_Data/Localization/Upgrade/EN_410_420.nopres.xml (hunk 4 above)
Comment (PatrykPorabik) [reply Grand.Web/App_Data/Localization/Upgrade/EN_410_420.nopres.xml]: And this one should be deleted...

**Pull request closed without merging**